docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,294 @@
+
+Cons: Requirement 1.b, 1.c are not met.
+
+#### Option 2
+
+Keep all collations defined in TiDB as what they were, define a series of new collations that are actually compatible with MySQL collations. For example, we can create a new `tidb_utf8_mb4_general_ci` that is the same as `utf8mb4_general_ci` in MySQL. When TiDB users want the "real" collations, they can modify their `CREATE TABLE` statements to use the new ones.
+
+Pros: Requirement 1.a, 1.b, 2 and 3 are met.
+
+Cons: Requirement 1.c and 4 are not met, the namings are also confusing: collations that are named from MySQL are different from MySQL, but the ones named different from MySQL behave the same as MySQL.
+
+#### Option 3
+
+Add no new collation, tag old collations with syntax comment. For a upgraded TiDB cluster, add comment like "`/* AS_BINARY_COLLATION */`" for the old collations; all collations with "`/* AS_BINARY_COLLATION */`" comment are treated as the old ones.
+
+Pros: Requirement 1.a, 2, 3 and 4 are met.
+
+Cons: Requirement 1.b and 1.c are not met.
+
+#### Option 4
+
+Only TiDB clusters that initially boostraped with the new TiDB version are allowed to enable the new collations. For old TiDB clusters, everything remains unchanged after the upgrade.
+
+We can also provide a configuration entry for the users to choose between old/new collations when deploying a new clusters.
+
+Pros: Requirement 1 and 2 are met.
+
+Cons: Since existing TiDB cluster can't get new collations enabled, requirement 3 is eliminated; Requirement 4 is not met.
+
+#### Bug-fixing
+
+We should also expect potential bugs in implementations of collations and get prepared for fixing them in the newer version of TiDB. However, this is not trivial since the collation affects the on-storage data, so a online data reorganization is required, which is a known limitation of TiDB. For now, another proposal is planning to support reorganizing data(the "online column type change"), we can expect it to be finished in the near future. Based on that, a possible solution for collation bug-fixing can be:
+
+  * Store a new version for the collations in the metadata of the column, it can be named as "CollationVersion" just like the current [ColumnInfoVersion](https://github.com/pingcap/parser/blob/b27aebeca4ba3fd938900fb57c0ea45c55d3753a/model/model.go#L66). This version is used to tracking the on-storage data generated by the collation.
+  * When we want to fix the bugs, related column index can be reorganized by a TiDB-specific DDL SQL like `ALTER TABLE ... TIDB UPGRADE COLLATION`. When the job is done, collation version is updated to the latest one.
+
+## Implementation
+
+### Collations for 8-bit charsets
+
+The implementation can be simply per-character mapping, same as MySQL.
+
+### Collations for Unicode charsets
+
+The following features of the general collation algorithm will be supported:
+
+  * Primary Weight i.e. character
+  * Secondary Weight i.e. accent
+  * Tertiary Weight i.e. case
+  * PAD / NOPAD
```

Comment:
According to https://godoc.org/golang.org/x/text/collate#Option there is no option for PAD / NOPAD. Do we need to implement the padding behaviour it by our own?

Reply:
Yes, I think so.